python/ray/util/client/server/logservicer.py

```diff
@@ -81,8 +82,25 @@ def log_status_change_thread(log_queue, request_iterator):
 
 
 class LogstreamServicer(ray_client_pb2_grpc.RayletLogStreamerServicer):
+    def __init__(self):
+        super().__init__()
+        self.num_clients = 0
+        self.client_lock = threading.Lock()
+
     def Logstream(self, request_iterator, context):
-        logger.info("New logs connection")
+        initialized = False
+        with self.client_lock:
```

Comment:
Do we need this here? won't the check in dataservicer suffice?

Reply:
Yeah, we do -- unfortunately without this check, the code will hang. 

The stacktrace is here for the client server process:

```

Thread 0x11B9ECDC0 (idle): "MainThread"
    main (ray/util/client/server/server.py:554)
    <module> (ray/util/client/server/__main__.py:3)
    _run_code (runpy.py:85)
    _run_module_as_main (runpy.py:193)
Thread 0x70000A21F000 (idle): "Thread-1"
    _serve (grpc/_server.py:871)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000A722000 (idle): "ThreadPoolExecutor-0_0"
    wait (threading.py:296)
    get (queue.py:170)
    Logstream (ray/util/client/server/logservicer.py:112)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000AC25000 (idle): "ThreadPoolExecutor-0_1"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    Datapath (ray/util/client/server/dataservicer.py:58)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000B128000 (idle): "ThreadPoolExecutor-0_2"
    wait (threading.py:296)
    get (queue.py:170)
    Logstream (ray/util/client/server/logservicer.py:112)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000CFD8000 (idle): "ThreadPoolExecutor-0_3"
    wait (threading.py:296)
    get (queue.py:170)
    Logstream (ray/util/client/server/logservicer.py:112)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000D4DB000 (idle): "ThreadPoolExecutor-0_4"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    Datapath (ray/util/client/server/dataservicer.py:58)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000D9DE000 (idle): "ThreadPoolExecutor-0_5"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    Datapath (ray/util/client/server/dataservicer.py:58)
    _take_response_from_response_iterator (grpc/_server.py:453)
    _send_message_callback_to_blocking_iterator_adapter (grpc/_server.py:607)
    _stream_response_in_pool (grpc/_server.py:593)
    run (concurrent/futures/thread.py:57)
    _worker (concurrent/futures/thread.py:80)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000B62B000 (idle): "Thread-3"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    log_status_change_thread (ray/util/client/server/logservicer.py:61)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000BB2E000 (idle): "Thread-4"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    log_status_change_thread (ray/util/client/server/logservicer.py:61)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
Thread 0x70000C2C0000 (idle): "Thread-5"
    wait (threading.py:296)
    _next (grpc/_server.py:373)
    __next__ (grpc/_server.py:382)
    log_status_change_thread (ray/util/client/server/logservicer.py:61)
    run (threading.py:870)
    _bootstrap_inner (threading.py:926)
    _bootstrap (threading.py:890)
```